### PR TITLE
added contributor email in admin councillor contribution#show page 

### LINKED ISSUE #1217: Administrators not notified of new councillor data contributions
So they know contributions are awaiting their review and merge.

### LINKED ISSUE #1218: Contributor's email address not shown to admin
So they can thank them :smile:

### PR DESCRIPTION
**Pull request by hisayohorie**
This is to fix #1218 

I updated the admin councillor contribution #show page to look like this, so admin can immediately see the contact information of the contributor without going to contributor#show page 

screen shot 2017-08-16 at 11 31 55 pm

There can be some layout polishing. Please let me know what you think @henare @equivalentideas

### CODE REVIEW

**Comment by equivalentideas** [comment]
Nice @hisayohorie . I agree the layout could be refined a bit. A more compact way to do it might be to have the 'Councillor Contribution Details' box say:

> * Authority: Marrickville Council
> * Contributor: Hisayo (hisayo@email.com)

Then you don't need the 'Contributor' heading or table at all. What do you think @hisayohorie ?

**Comment by hisayohorie** [comment]
I created a partial for the customized view of the contributor's name&email and authority the suggested councillor belong to. To avoid the duplication, previous contributor table with the "Contributor" heading is removed.

A line of test is added to check to see if the contributor's email is displayed
Also I changed the expectation of display from "Contributor" from "Contributor Name" to adjust the change in the table heading.

It is displayed as below:

screen shot 2017-08-17 at 1 09 22 pm

Please let me know what you think @henare @equivalentideas !

**Comment by hisayohorie** [comment]
I am having another Travis issue, which seems to be related to ActiveAdmin and/or Devise. Now investigating

**Comment by henare** [comment]
>This is to fix #1217

@hisayohorie I think you mean #1218 :smile:

**Comment by henare** [comment]
>I am having another Travis issue, which seems to be related to ActiveAdmin and/or Devise. Now investigating

@hisayohorie I get those failures locally so I don't think it's a Travis-specific issue :bug:

**Comment by hisayohorie** [comment]
@henare it wasn't! 
I looked into the file and `admin/contributor.rb` wasn't pushed 😅

File: app/admin/contributors.rb [hunk 1]
```diff
@@ -0,0 +1,11 @@
+ActiveAdmin.register Contributor do
```
Comment (henare) [inline app/admin/contributors.rb]: Why do we need this new admin page?

As the admin I can't think how a list of contributors is useful. I only need to see it in conjunction with their contribution.

File: app/admin/contributors.rb [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+    actions :index, :show
+
+    index do
+      column :name
+      column :email
+      column :id
```
Comment (henare) [inline app/admin/contributors.rb]: I definitely don't need to know the internal database ID :smile:

File: app/admin/councillor_contributions.rb [hunk 3]
```diff
@@ -10,10 +10,9 @@
   end
 
   show title: :created_at do
-    attributes_table do
-      row :contributor
-      row :id
-      row :authority
+
+    panel "Councillor Contribution Details" do
+      render partial: "details", locals: {contributor: resource.contributor, authority: resource.authority}
```
Comment (henare) [inline app/admin/councillor_contributions.rb]: If the contributor doesn't submit their contact details this page will now fail (the exception is `undefined method 'name' for nil:NilClass`.

File: app/views/admin/councillor_contributions/_details.html.arb [hunk 4]
```diff
@@ -0,0 +1,13 @@
+table do
+  tr do
+    th "Contributor Name"
+    th "Contributor Email"
+    th "Authority"
+  end
+  tr do
+    td a contributor.name, href:admin_contributor_path
+    td contributor.email
+    "#{contributor.email}"
```
Comment (henare) [inline app/views/admin/councillor_contributions/_details.html.arb]: What does this line do?

File: app/views/admin/councillor_contributions/_details.html.arb [hunk 5]
```diff
@@ -0,0 +1,13 @@
+table do
+  tr do
+    th "Contributor Name"
+    th "Contributor Email"
+    th "Authority"
+  end
+  tr do
+    td a contributor.name, href:admin_contributor_path
+    td contributor.email
```
Comment (henare) [inline app/views/admin/councillor_contributions/_details.html.arb]: A `mailto:` link could be useful to allow the admin to quickly email the contributor.

File: app/views/admin/councillor_contributions/_details.html.arb [hunk 6]
```diff
@@ -0,0 +1,13 @@
+table do
+  tr do
+    th "Contributor Name"
+    th "Contributor Email"
+    th "Authority"
+  end
+  tr do
+    td a contributor.name, href:admin_contributor_path
```
Comment (henare) [inline app/views/admin/councillor_contributions/_details.html.arb]: When I visit this link, what extra information do I get? This relates to my comment questioning if we need the new `app/admin/contributors.rb` admin section.

File: app/views/admin/councillor_contributions/_details.html.arb [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+    td contributor.email
+    "#{contributor.email}"
+    td a authority.full_name
```
Comment (henare) [inline app/views/admin/councillor_contributions/_details.html.arb]: This doesn't link to anything when I test locally. What should it be going to and why?

File: spec/features/admin_views_councillor_contribution_spec.rb [hunk 8]
```diff
@@ -51,10 +51,25 @@
 
     click_link "View"
 
-    expect(page).to have_content "Contributor Felix Chaung"
-    expect(page).to have_content "Authority Casey City Council"
+    expect(page).to have_content "Felix Chaung"
+    expect(page).to have_content "example@gmail"
```
Comment (henare) [inline spec/features/admin_views_councillor_contribution_spec.rb]: Probably best to use the whole email address so it's clear what we're actually looking for, i.e. not just part of the email address but the whole thing.

File: app/admin/contributors.rb (hunk 1 above)
Comment (hisayohorie) [inline app/admin/contributors.rb]: @henare Ok, will delete 👻

File: app/admin/councillor_contributions.rb (hunk 3 above)
Comment (hisayohorie) [inline app/admin/councillor_contributions.rb]: screen shot 2017-08-18 at 4 05 48 pm
 
added a test as well

File: app/views/admin/councillor_contributions/_details.html.arb (hunk 6 above)
Comment (hisayohorie) [inline app/views/admin/councillor_contributions/_details.html.arb]: I deleted the link as it is not needed. It was going to contributor detail show page.

File: spec/features/admin_views_councillor_contribution_spec.rb (hunk 8 above)
Comment (hisayohorie) [inline spec/features/admin_views_councillor_contribution_spec.rb]: Yes, I updated it to a complete email format. .com slipped off my eyes.

File: app/views/admin/councillor_contributions/_details.html.arb (hunk 4 above)
Comment (hisayohorie) [inline app/views/admin/councillor_contributions/_details.html.arb]: doesn't do anything 😅  deleted.

File: app/views/admin/councillor_contributions/_details.html.arb (hunk 7 above)
Comment (hisayohorie) [inline app/views/admin/councillor_contributions/_details.html.arb]: It suppose to take it to authority#show page. I fixed it, but if it's not necessarily I will just make it to a text.

File: app/views/admin/councillor_contributions/_details.html.arb (hunk 5 above)
Comment (hisayohorie) [inline app/views/admin/councillor_contributions/_details.html.arb]: Added 🚀

**Comment by hisayohorie** [comment]
I changed the layout as below

screen shot 2017-08-18 at 5 34 52 pm

The title `:created_at` is now `:authority_name`.
To enable this, I created `authority_name` method in councillor_contribution model, which I am not sure if this is the best way. 

Also, now the title is authority name, is the column :authority necessary?  @henare @equivalentideas

**Comment by hisayohorie** [comment]
I accidentally included the fix for #1231 here. Sorry 😞 

The title `:created_at` is now `:authority_name`.
To enable this, I created `authority_name` method in councillor_contribution model, which I am not sure if this is the best way. 
Also, now the title is authority name, is the column :authority necessary? 

@henare @equivalentideas

**Comment by equivalentideas** [comment]
> The title :created_at is now :authority_name.
> To enable this, I created authority_name method in councillor_contribution model, which I am not sure if this is the best way.

@hisayohorie I recon a useful title would be "Councillor Contribution for Marrickville Council, August 23, 2017".

That tells the admin what they're looking at, a councillor contribution, what council it's for, and when it's from. Lots of useful information.

> Also, now the title is authority name, is the column :authority necessary?

Nah, I recon if the authority name is in the title then no need to repeat it.

**Comment by hisayohorie** [comment]
@equivalentideas As you suggested, I made a change to the title to show the date, also I removed the authority column since it is duplicating information. 

screen shot 2017-08-23 at 4 21 18 pm

File: spec/features/admin_views_councillor_contribution_spec.rb [hunk 9]
```diff
@@ -44,17 +44,68 @@
...
+
+    it "successfully" do
+      sign_in_as_admin
+
+      click_link "Councillor Contributions"
+
+      click_link "View"
+
+      expect(page).to have_content "Anonymous Contributor"
+      expect(page).to have_content "Casey City Council"
```
Comment (equivalentideas) [inline spec/features/admin_views_councillor_contribution_spec.rb]: Does this test need to cover all this other information? It reads to me like the purpose of this test is to make sure it says 'Anonymous contributor' when the contributor hasn't added their details?

This points to a good potential refactor here to me. The logic of how to display the name of the person is currently in the view, but it should be extracted to the model or a view helper—I think the model probably. Then we could also just cover the logic in a unit test. Could be a good pairing session?

**Comment by equivalentideas** [comment]
@hisayohorie this is looking good 👍 After our impromptu pairing session this morning I did end up doing a little more refactor to extract the `CouncillorContribution#attribution` method and add tests so we could remove that big integration test.

I also inlined one method you extracted, you can read about that here https://github.com/openaustralia/planningalerts/pull/1226/commits/e8d0a65d8662f26e74377616aac8452a74ab8fda There's a bit of the thinking on the refactors on the commit messages.

I'll merge and deploy!

**Comment by equivalentideas** [comment]
Deployed!

**Pull request merged**